lua/entities/kleiner/init.lua

```diff
@@ -0,0 +1,911 @@
+    self.TargetBlacklist = self.TargetBlacklist or {}
+    self.TargetBlacklist[target] = CurTime() + 5
+end
+
+function ENT:RunBehaviour()
+    -- This function is called when the entity is first spawned, it acts as a giant loop that will run as long as the NPC exists
+    while (true) do
+        if (self.loco:IsStuck()) then
+            self:HandleStuck()
+        end
+
+        if (self.NeedsTarget) then
+            self:FindTarget()
+            self.NeedsTarget = nil
+        end
+
+        if (IsValid(self:GetTarget()) and self:CanBeTarget(self:GetTarget())) then
+            self.loco:SetDesiredSpeed((self.Suicidal and 350) or (self:GetBased() and 500) or 200)
+            self.loco:SetAcceleration(self.Suicidal and 400 or 200)
+            self.loco:SetDeceleration(self.Suicidal and 400 or 500)
+            self.loco:SetJumpHeight(100)
+
+            if (not self:NearTarget()) then
+                local result = self:ChaseTarget()
+
+                --if chasing the target fails somehow, its probably a wise assumption that its redundant to keep trying.
+                if (result == "failed") then
+                    self:IgnoreTarget(self:GetTarget())
+                    self:ResetBehavior()
+                end
+            end
+
+            if (self:NearTarget()) then
+                if (not self:IsTalking()) then
+                    self:SayStuff()
+                end
+
+                self.loco:FaceTowards(self:GetTarget():GetPos())
+
+                if (not self.shifted) then
+                    local result = self:MoveToPos(self:GetTarget():GetPos() + VectorRand() * Vector(1, 1, 0):GetNormalized() * math.Rand(80, 150), {
+                        maxage = 50
+                    })
+
+                    if (result ~= "ok") then
+                        self:ResetBehavior()
+                    end
+
+                    --cheap method of keeping kleiners with similar targets from clumping together. i figure it's cheaper than some kind of avoidance.
+                    self.shifted = true
+                end
+
+                coroutine.wait(0.15)
+            else
+                self.shifted = nil
+            end
+
+            if (self.ManualTarget == nil and math.random(1, 100) == 1) then
+                self:ResetBehavior()
+            end
+        else
+            -- no target, so we wander.
+            self.loco:SetDesiredSpeed(self.Suicidal and 350 or 200)
+            self.loco:SetAcceleration(self.Suicidal and 400 or 200)
+            self.loco:SetDeceleration(self.Suicidal and 900 or 900)
+
+            if (math.random(1, 50) == 1) then
+                self:ResetBehavior()
+            else
+                local wanderpos
+
+                if (self.WanderForcePos) then
+                    wanderpos = self.WanderForcePos + VectorRand() * Vector(1, 1, 0):GetNormalized() * math.Rand(80, 150)
+                end
+
+                wanderpos = wanderpos or self:FindSpot("random", {
```

Comment:
This seems to be the main lag causer, looking at https://github.com/Facepunch/garrysmod/blob/be251723824643351cb88a969818425d1ddf42b3/garrysmod/gamemodes/base/entities/entities/base_nextbot/sv_nextbot.lua it seems to essentially find a bunch of nearby areas, then make a path to each area (!!!). We should probalby use a much lower radius so it finds less areas, and it can find its way to distant areas after multiple wanders. It might also work to override ent:FindSpots to not calculate a path and just use the vector:Distance() instead.